### PR TITLE
[RG-194, RG-195] Fill Placement/Routing messages.

### PR DESCRIPTION
**Pull request by TarasSydor**
Given PR implements RG-194 and RG-195 features - Placement and Routing messages shown in Messages tab.

> ### Describe the technical details
Parsing is done line by line. Some logic is placed to Abstract level so that it can be reused by other parsers. Common code parses certain section from a log file, highlighting also some key messages inside. The only requirement is - end of section line starts with the same string as its beginning.

One more important change - TaskMessage structure, which allows multi-level trees and setting a message severity.
image

Some aspects of behavior:
- 4 categories added - ERROR, WARNING, INFO, NONE (used for sections and task items)
- In case no warnings or errors found, no message is shown
- In case multiple warnings/errors were found, double clicking on a corresponding message navigates to the first occurred message/error:
image

### CODE REVIEW

File: src/Compiler/Reports/ITaskReportManager.h [hunk 1]
```diff
@@ -26,6 +26,20 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
 namespace FOEDAG {
 
+struct TaskMessage {
+  enum class MessageSeverity {
```
Comment (KochynVolodymyr) [inline src/Compiler/Reports/ITaskReportManager.h:30]: I think enum class could be moved outside the struct to avoid long access string.

**Review by KochynVolodymyr — APPROVED** [review]

**Review by skyler-rs — APPROVED** [review]

File: src/Compiler/Reports/ITaskReportManager.h (hunk 1 above)
Comment (TarasSydor) [reply src/Compiler/Reports/ITaskReportManager.h]: agree. fixed

**Comment by TarasSydor** [comment]
After team validation of Messages tab we came up with some remarks to existing behavior, but it will come in a separate PR.
- Timing report should be created for Placement;
- Messages tab should contain all errors and messages (hidden under collapsible parent item);
- If no errors/warnings found, ''0 errors found" item should not be added;
- Warnings should be grouped. Info messages get sorted by line number;
- Nice to have: collapse/expand all, configuration to hide/show warnings/errors.

**Comment by TarasSydor** [comment]
Changes added by latest commit:
- Messages tab now contains all errors/warnings. Actual messages are shown without an icon within collapsible parent item;
- Messages are sorted by the line number. Warning/Error parent items line number is determined by the first warning/error in a group;
- Synthesis errors/warnings are reported and grouped in a similar to others way: parent item, encapsulating warnings/errors.
- Fixed the issue with Placement resource usage report.

File: src/Compiler/Reports/AbstractReportManager.cpp [hunk 2]
```diff
@@ -106,16 +104,18 @@ int AbstractReportManager::parseErrorWarningSection(QTextStream &in, int lineNr,
 
   QString line;
   // Store line numbers in set to always know which one was first
-  std::set<int> wrnLines, errLines;
+  std::map<int, QString> warnings, errors;
```
Comment (KochynVolodymyr) [inline src/Compiler/Reports/AbstractReportManager.cpp:107]: Are you sure you didn't reorder the elements by using map/set? As we discussed we should keel the order of elements.

Comment (TarasSydor) [reply src/Compiler/Reports/AbstractReportManager.cpp:107]: Yes. both std::set and std::map use std::less<Key> comparator by default.
So we end up with the earliest message/error going first in the container.

Comment (TarasSydor) [reply src/Compiler/Reports/AbstractReportManager.cpp:117]: Recent commit implements following grouping:
image

Errs/warnings, that aren't split by any Info message, are grouped within a single item. For standalone warnings/errors no grouping is made.

File: src/Compiler/Reports/AbstractReportManager.cpp [hunk 3]
```diff
@@ -128,28 +128,43 @@ int AbstractReportManager::parseErrorWarningSection(QTextStream &in, int lineNr,
...
 
+TaskMessage AbstractReportManager::createWarningErrorItem(
+    MessageSeverity severity, const MessagesLines &msgs) const {
+  auto itemName = severity == MessageSeverity::ERROR_MESSAGE
+                      ? QString("Errors")
+                      : QString("Warnings");
+  auto itemLine = msgs.begin()->first;
+
+  auto itemMsg =
+      QString("%1 %2 found").arg(QString::number(msgs.size()), itemName);
```
Comment (KochynVolodymyr) [inline src/Compiler/Reports/AbstractReportManager.cpp:196]: 1 Errors found would be a typo

**Pull request merged**